File: 02 tests/ex4_count_chars.py
#!/usr/bin/env python3

""" ex4_count_chars.py

Count a's in a 7 chars random string

:author:	wolf
:created:	2025.05.22
"""


import time
import random
import string

iterations = 3
timing_results = []

# define the test function
# --- function to be timed ---


def python_char_count(n):
    text = ''.join(random.choice(string.ascii_lowercase) for _ in range(n))
    return sum(1 for c in text if c == 'a')


for i in range(iterations):
    start = time.perf_counter()

    # call the test function
    python_char_count(10**7)

    end = time.perf_counter()

    timing_results.append(end - start)


print(f"Elapsed time:  {sum(timing_results)/iterations:.4f} seconds")

File: 02 tests/ex4_count_chars.js
// run deno ex4_count_chars.js | node ex4_count_chars.js

/* ex4_count_chars.js

Count a's in a 10^7 chars random string

:author:	wolf
:created:	2025.05.22
*/


const iterations = 3;
const timing_results = [];

// define the test function
// --- function to be timed-- -

// Array.from does not work with arrays size > 10^7
// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Invalid_array_length

function jsCharCount(n) {
    const chars = "abcdefghijklmnopqrstuvwxyz";
    const text = Array.from({ length: n }, () => chars[Math.floor(Math.random() * chars.length)]).join('');
    return text.split('').filter(c => c === 'a').length;
}


for (let i = 0; i < iterations; i++) {
    const start = performance.now();

    // call the test function
    jsCharCount(10**7);

    const end = performance.now();

    timing_results.push(end - start);
}


console.log("Elapsed time: ",
    timing_results.reduce((a, b) => a + b) / iterations / 1000, "seconds");